File: javascript/beer-song/beer.js
"use strict";

var Beer = {};

Beer.verse = function(count) {
    var countString = function(count) {
        if (count === 0)
            return "no more bottles";
        if (count === 1)
            return "1 bottle"
        return count.toString() + " bottles";
    };

    var bottleString = function (count) {
        if (count === 1)
            return "it";
        return "one";
    };

    if (count === 0)
        return "No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n";

    return countString(count) + " of beer on the wall, " + countString(count) + 
        " of beer.\nTake " + bottleString(count) + " down and pass it around, " + 
        countString(count - 1) + " of beer on the wall.\n"
};

Beer.sing = function(begin, end) {
    var result = "";
    end = end || 0; 
    for(; begin>=end ; begin--) {
        result = result + Beer.verse(begin) + (begin === end ? "" : "\n");
    }
    return result;
};


module.exports = Beer;

File: clojure/beer-song/beer.clj
(ns beer
  (:require [clojure.string :as string]))

(defn- count-string [count]
  (cond 
    (zero? count) "no more bottles"
    (= count 1)   "1 bottle"
    :else         (str count " bottles")))

(defn- singular-string [count]
  (if (= count 1) 
    "it" 
    "one"))

(defn- first-strophe [count]
  (string/capitalize 
    (let [n-bottles (count-string count)]
      (str n-bottles " of beer on the wall, " n-bottles " of beer."))))

(defn- second-strophe [count]
  (if (zero? count)
    "Go to the store and buy some more, 99 bottles of beer on the wall.\n"
    (let [singular (singular-string count) 
          n-bottles (count-string (dec count))]
      (str "Take " singular " down and pass it around, " n-bottles " of beer on the wall.\n"))))
  
(defn verse [count]
  (str (first-strophe count) "\n" (second-strophe count)))

(defn sing 
  ([begin] (sing begin 0))
  ([begin end]
    (str (string/join "\n" (map verse (range begin (dec end) -1))) "\n")))